### PR TITLE
Add type hinting to public-facing API  in pandas/core/generic.py #26792

### PR DESCRIPTION
**Pull request by tobyjamez**
- [x] Part of #26792
- [x] tests passed
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`

### CODE REVIEW

File: pandas/core/generic.py [hunk 1]
```diff
@@ -728,7 +730,7 @@ def _set_axis(self, axis, labels):
         self._data.set_axis(axis, labels)
         self._clear_item_cache()
 
-    def transpose(self, *args, **kwargs):
+    def transpose(self, *args, **kwargs) -> "NDFrame":
```
Comment (jorisvandenbossche) [inline pandas/core/generic.py]: If we start adding type hints to public functions (maybe we already have other places where we do this, but just noticing it here): we should think about what to do with "NDFrame". 
This is not a "public term", and IMO no user should see this.

Should it be replaced with a union of Series and Frame?

**Comment by jbrockmendel** [comment]
you need to run `isort pandas/core/generic.py`

File: pandas/core/generic.py [hunk 2]
```diff
@@ -553,7 +555,7 @@ def _stat_axis(self):
         return getattr(self, self._stat_axis_name)
 
     @property
-    def shape(self):
+    def shape(self) -> Tuple[int]:
```
Comment (jbrockmendel) [inline pandas/core/generic.py]: should this be `Tuple[int, ...]`?

File: pandas/core/generic.py [hunk 3]
```diff
@@ -772,7 +774,7 @@ def transpose(self, *args, **kwargs):
         nv.validate_transpose(tuple(), kwargs)
         return self._constructor(new_values, **new_axes).__finalize__(self)
 
-    def swapaxes(self, axis1, axis2, copy=True):
+    def swapaxes(self, axis1: int, axis2: int, copy=True) -> "NDFrame":
```
Comment (jbrockmendel) [inline pandas/core/generic.py]: `int` here isnt quite right.  I think its `pandas._typing.Axis`

File: pandas/core/generic.py (hunk 1 above)
Comment (jbrockmendel) [reply pandas/core/generic.py]: @simonjayhawkins should this by `pd._typing._T`?

Comment (simonjayhawkins) [reply pandas/core/generic.py]: > @simonjayhawkins should this by `pd._typing._T`?

`pd._typing._T` is for unbound types (normally objects in a generic container). we should use `TypeVar("FrameOrSeries", bound="NDFrame")` in `core.generic`

this is currently `pd._typing.FrameOrSeries`. some potentially relevant discussion in #29480 

> This is not a "public term", and IMO no user should see this.

that's a reasonable agrument and perhaps also relevant to #29480

**Comment by jreback** [comment]
closing as stale, some of this has been done, welcome for PRs to continue

**Pull request closed without merging**